### DISCUSSION TITLE
Does CDI work with extensions in the new Keycloak-X

### DISCUSSION THREAD

**Discussion by tlann** [keycloak/keycloak#12158] [Archive]
I'm updating our token mapper to work with the new Keycloak-X.   I have been unable to inject and entitymanager or our repository.

I have been able to programmatically get a entityManager because of some code I saw from [pedroigor](https://github.com/pedroigor) 
It involves using `em = keycloakSession.getProvider(JpaConnectionProvider.class, "users").getEntityManager();`

Does the quarkus version of CDI work in extensions?  I have been unable to get it to work.

**Comment by pedroigor** [comment]
Your custom providers are not registered as beans when the server is bootstrapping.

What you can do though is to use something like this https://github.com/keycloak/keycloak/blob/b5a5d68dbc06390d26c1c0b628e4544f718ebd3a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/database/jpa/QuarkusJpaConnectionProviderFactory.java#L195.

There we are using `Arc` directly but better is to use standard CDI like this https://github.com/keycloak/keycloak/blob/b4c837f14810d09f016ebf37d346339957dffaa3/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/transaction/QuarkusJtaTransactionManagerLookup.java#L39.

We would like to improve this part and make providers fully integrated with Quarkus. For instance, your mapper will be installed using a specific annotation and able leverage CDI.

We could potentially drop the need for implementing factories too so you only need to implement the provider interface. if you need to initialize a provider at init phase, then you would probably do that listening for CDI events.

However, we should try as much as possible to not introduce 3rd party dependencies to your providers but just use what Keycloak provides (yer, I know is not much). For instance, I would get the `EntityManagerFactory` from the `KeycloakSession` as you previously mentioned.

**Reply by tlann** [reply]
@pedroigor Thanks for answering.  
It seems like I used the acronym CDI wrong.  
I was talking more of the @Inject annotation or its analogs for quarkus for the EntityManager and other Beans like a UserExtraDetailsRepo

I tried looking at the Quarkus documentation and using a @Produces annotation for the above beans but couldn't get those to work.

**Reply by arrudeboy** [reply]
Hi there! I was dealing with the same issue. I would like to use jakarta **@Inject** annotation in order to inject some beans which are treated as singleton. Example:

```
@ApplicationScoped
public class ConsentAccountsRequiredAction implements RequiredActionProvider, RequiredActionFactory {

    @Inject
    AccountService accountService;

    @Override
    public void requiredActionChallenge(RequiredActionContext context) {
        var user = context.getUser();
        var customerId = user.getFirstAttribute("customerId");
        var accounts = accountService.getCbuList(customerId);
        ...
    }
```


But **accountService** object always remains null. 

I know that there're things that I am doing wrong. I would appreciate if anyone can provide an example of configuring beans within a SPI (in my case is a RequiredAction provider). 

Thanks in advance!

**Reply by codespearhead** [reply]
To the two questions above, the original answer explains that Keycloak's SPI doesn’t support CDI injection yet, which is why you should use factories for dependency management.

***

Related: https://github.com/keycloak/keycloak/discussions/22687#discussioncomment-8484254

**Reply by arrudeboy** [reply]
Thank you very much for your reply! 

So I'll have to make it works by implementing factories, or let's say "manual singletons" instantiated from the SPI factory (config) class.

**Reply by codespearhead** [reply]
Exactly.

**Reply by codespearhead** [reply]
In the meantime, this project [01][02] might help depending on your needs.

[01] https://github.com/keycloak/keycloak/discussions/13504
[02] https://github.com/keycloak/keycloak/discussions/13504#discussioncomment-11553981
